<?xml version="1.0" encoding="utf-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="Docutils 0.11: http://docutils.sourceforge.net/" />
<title>PubMed2Go</title>
<style type="text/css">

/*
:Author: David Goodger (goodger@python.org)
:Id: $Id: html4css1.css 7614 2013-02-21 15:55:51Z milde $
:Copyright: This stylesheet has been placed in the public domain.

Default cascading style sheet for the HTML output of Docutils.

See http://docutils.sf.net/docs/howto/html-stylesheets.html for how to
customize this style sheet.
*/

/* used to remove borders from tables and images */
.borderless, table.borderless td, table.borderless th {
  border: 0 }

table.borderless td, table.borderless th {
  /* Override padding for "table.docutils td" with "! important".
     The right padding separates the table cells. */
  padding: 0 0.5em 0 0 ! important }

.first {
  /* Override more specific margin styles with "! important". */
  margin-top: 0 ! important }

.last, .with-subtitle {
  margin-bottom: 0 ! important }

.hidden {
  display: none }

a.toc-backref {
  text-decoration: none ;
  color: black }

blockquote.epigraph {
  margin: 2em 5em ; }

dl.docutils dd {
  margin-bottom: 0.5em }

object[type="image/svg+xml"], object[type="application/x-shockwave-flash"] {
  overflow: hidden;
}

/* Uncomment (and remove this text!) to get bold-faced definition list terms
dl.docutils dt {
  font-weight: bold }
*/

div.abstract {
  margin: 2em 5em }

div.abstract p.topic-title {
  font-weight: bold ;
  text-align: center }

div.admonition, div.attention, div.caution, div.danger, div.error,
div.hint, div.important, div.note, div.tip, div.warning {
  margin: 2em ;
  border: medium outset ;
  padding: 1em }

div.admonition p.admonition-title, div.hint p.admonition-title,
div.important p.admonition-title, div.note p.admonition-title,
div.tip p.admonition-title {
  font-weight: bold ;
  font-family: sans-serif }

div.attention p.admonition-title, div.caution p.admonition-title,
div.danger p.admonition-title, div.error p.admonition-title,
div.warning p.admonition-title, .code .error {
  color: red ;
  font-weight: bold ;
  font-family: sans-serif }

/* Uncomment (and remove this text!) to get reduced vertical space in
   compound paragraphs.
div.compound .compound-first, div.compound .compound-middle {
  margin-bottom: 0.5em }

div.compound .compound-last, div.compound .compound-middle {
  margin-top: 0.5em }
*/

div.dedication {
  margin: 2em 5em ;
  text-align: center ;
  font-style: italic }

div.dedication p.topic-title {
  font-weight: bold ;
  font-style: normal }

div.figure {
  margin-left: 2em ;
  margin-right: 2em }

div.footer, div.header {
  clear: both;
  font-size: smaller }

div.line-block {
  display: block ;
  margin-top: 1em ;
  margin-bottom: 1em }

div.line-block div.line-block {
  margin-top: 0 ;
  margin-bottom: 0 ;
  margin-left: 1.5em }

div.sidebar {
  margin: 0 0 0.5em 1em ;
  border: medium outset ;
  padding: 1em ;
  background-color: #ffffee ;
  width: 40% ;
  float: right ;
  clear: right }

div.sidebar p.rubric {
  font-family: sans-serif ;
  font-size: medium }

div.system-messages {
  margin: 5em }

div.system-messages h1 {
  color: red }

div.system-message {
  border: medium outset ;
  padding: 1em }

div.system-message p.system-message-title {
  color: red ;
  font-weight: bold }

div.topic {
  margin: 2em }

h1.section-subtitle, h2.section-subtitle, h3.section-subtitle,
h4.section-subtitle, h5.section-subtitle, h6.section-subtitle {
  margin-top: 0.4em }

h1.title {
  text-align: center }

h2.subtitle {
  text-align: center }

hr.docutils {
  width: 75% }

img.align-left, .figure.align-left, object.align-left {
  clear: left ;
  float: left ;
  margin-right: 1em }

img.align-right, .figure.align-right, object.align-right {
  clear: right ;
  float: right ;
  margin-left: 1em }

img.align-center, .figure.align-center, object.align-center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}

.align-left {
  text-align: left }

.align-center {
  clear: both ;
  text-align: center }

.align-right {
  text-align: right }

/* reset inner alignment in figures */
div.align-right {
  text-align: inherit }

/* div.align-center * { */
/*   text-align: left } */

ol.simple, ul.simple {
  margin-bottom: 1em }

ol.arabic {
  list-style: decimal }

ol.loweralpha {
  list-style: lower-alpha }

ol.upperalpha {
  list-style: upper-alpha }

ol.lowerroman {
  list-style: lower-roman }

ol.upperroman {
  list-style: upper-roman }

p.attribution {
  text-align: right ;
  margin-left: 50% }

p.caption {
  font-style: italic }

p.credits {
  font-style: italic ;
  font-size: smaller }

p.label {
  white-space: nowrap }

p.rubric {
  font-weight: bold ;
  font-size: larger ;
  color: maroon ;
  text-align: center }

p.sidebar-title {
  font-family: sans-serif ;
  font-weight: bold ;
  font-size: larger }

p.sidebar-subtitle {
  font-family: sans-serif ;
  font-weight: bold }

p.topic-title {
  font-weight: bold }

pre.address {
  margin-bottom: 0 ;
  margin-top: 0 ;
  font: inherit }

pre.literal-block, pre.doctest-block, pre.math, pre.code {
  margin-left: 2em ;
  margin-right: 2em }

pre.code .ln { color: grey; } /* line numbers */
pre.code, code { background-color: #eeeeee }
pre.code .comment, code .comment { color: #5C6576 }
pre.code .keyword, code .keyword { color: #3B0D06; font-weight: bold }
pre.code .literal.string, code .literal.string { color: #0C5404 }
pre.code .name.builtin, code .name.builtin { color: #352B84 }
pre.code .deleted, code .deleted { background-color: #DEB0A1}
pre.code .inserted, code .inserted { background-color: #A3D289}

span.classifier {
  font-family: sans-serif ;
  font-style: oblique }

span.classifier-delimiter {
  font-family: sans-serif ;
  font-weight: bold }

span.interpreted {
  font-family: sans-serif }

span.option {
  white-space: nowrap }

span.pre {
  white-space: pre }

span.problematic {
  color: red }

span.section-subtitle {
  /* font-size relative to parent (h1..h6 element) */
  font-size: 80% }

table.citation {
  border-left: solid 1px gray;
  margin-left: 1px }

table.docinfo {
  margin: 2em 4em }

table.docutils {
  margin-top: 0.5em ;
  margin-bottom: 0.5em }

table.footnote {
  border-left: solid 1px black;
  margin-left: 1px }

table.docutils td, table.docutils th,
table.docinfo td, table.docinfo th {
  padding-left: 0.5em ;
  padding-right: 0.5em ;
  vertical-align: top }

table.docutils th.field-name, table.docinfo th.docinfo-name {
  font-weight: bold ;
  text-align: left ;
  white-space: nowrap ;
  padding-left: 0 }

/* "booktabs" style (no vertical lines) */
table.docutils.booktabs {
  border: 0px;
  border-top: 2px solid;
  border-bottom: 2px solid;
  border-collapse: collapse;
}
table.docutils.booktabs * {
  border: 0px;
}
table.docutils.booktabs th {
  border-bottom: thin solid;
  text-align: left;
}

h1 tt.docutils, h2 tt.docutils, h3 tt.docutils,
h4 tt.docutils, h5 tt.docutils, h6 tt.docutils {
  font-size: 100% }

ul.auto-toc {
  list-style-type: none }

</style>
</head>
<body>
<div class="document" id="pubmed2go">
<h1 class="title">PubMed2Go</h1>

<div class="section" id="introduction">
<h1>Introduction</h1>
<ul class="simple">
<li>PubMed2Go processes XML files that can be downloaded from NCBI and builds up a PostgreSQL database containing all abstracts of the data set (no full texts).</li>
<li>Furthermore, PubMed2Go generates a full text index with Xapian such that titles, abstracts, keywords, MeSH terms and substances can be queried with search terms.</li>
<li>It is scalable to your system requirements using multiprocessing and can be modified easily to your personal needs.</li>
<li>This documentation refers to a text mining example referring to the disease pancreatic cancer. It contains many examples how to use PostgreSQL and Xapian as well as connecting both with small Python programmes. The default values in the scripts refer to the parameters given here, but all of them can be changed by the user.</li>
<li>Pancreatic cancer is one of the most dangerous cancer types, because the only way to cure a patient is a surgery beside several therapeutic strategies that could not significantly increase survival rates [Pancreatic cancer: from state-of-the-art treatments to promising novel therapies. Garrido-Laguna I, Hidalgo M. Nat Rev Clin Oncol. 2015 Mar 31. doi: 10.1038/nrclinonc.2015.53.].</li>
<li>Considering text mining as an approach to have a closer look at related genes/proteins, chemical compounds, and diseases in documents that are relevant for pancreatic cancer shows a need for an easy-to-use software bridging the gap between processing data sets from PubMed effectively on a local machine and using the range of tools offered for natural language processing, e.g. by the BioCreative community. PubMed2Go offers such a solution.</li>
<li>If the examples from this documentations are used, there will be around 745 MB of disk space needed. There are no other hardware requirements.</li>
<li>Start by copying the whole project folder from GitHub to your local disk.</li>
</ul>
</div>
<div class="section" id="download-a-data-set">
<h1>Download a Data Set</h1>
<ul>
<li><p class="first">Type in the following URL into your browser and search for &quot;pancreatic cancer&quot;:</p>
<blockquote>
<ul>
<li><p class="first"><a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/">http://www.ncbi.nlm.nih.gov/pubmed/</a></p>
</li>
<li><p class="first">The quote signs are important for the exact search.</p>
</li>
<li><p class="first">Click on &quot;Send to:&quot; and &quot;File&quot; and select &quot;XML&quot;.</p>
</li>
<li><p class="first">16th April 2015: 23258 PubMed-IDs</p>
<blockquote>
<ul class="simple">
<li>Some PubMed-IDs might change over time. Even for the given example list of PubMed-IDs for this documentation in &quot;data/pubmed_result.txt&quot; it is possible, that you receive another number of downloaded publications in your XML files as well as different outcomes in the ongoing analyes.</li>
</ul>
</blockquote>
</li>
<li><p class="first">The download of around 272 MB can take up to one hour depending on the time of day and internet connection.</p>
</li>
</ul>
</blockquote>
</li>
<li><p class="first">Another possibility is to download PubMed articles with EFetch:</p>
<blockquote>
<ul>
<li><p class="first">How to build a query:</p>
<blockquote>
<ul class="simple">
<li><a class="reference external" href="http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EFetch">http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.EFetch</a></li>
<li><a class="reference external" href="http://www.ncbi.nlm.nih.gov/books/NBK25501/">http://www.ncbi.nlm.nih.gov/books/NBK25501/</a></li>
</ul>
</blockquote>
</li>
<li><p class="first">You can also download only the PubMed-ID list from your browser by selecting &quot;PMID List&quot; instead of &quot;XML&quot; and then use EFetch (tested only on Ubuntu):</p>
<blockquote>
<ul class="simple">
<li>Change into the directory &quot;data&quot; in your command-line and type in &quot;python generate_efetch.py&quot;. It will build the script &quot;efetch.sh&quot; which generates XML files with 100 PubMed-IDs in each document.</li>
<li>The standard input filename is &quot;pubmed_result.txt&quot; and the block size of PubMed-IDs for one XML file is 100, but this can be changed. Type in &quot;python generate_efetch.py -h&quot; to show adjustable parameters.</li>
<li>If you want to specify the name of the directory where the XML files should be saved, use parameter &quot;-d&quot; and create this folder manually before.</li>
<li>The file &quot;data/pubmed_result.txt&quot; contains 23258 PubMed-IDs and the first 100 PubMed-IDs are saved in &quot;data/pancreatic_cancer_example/medline_00000000.xml&quot; as an example.</li>
<li>Make &quot;efetch.sh&quot; executable with &quot;chmod +x efetch.sh&quot; in the command-line and run it in a folder of your choice by typing in &quot;./efetch.sh&quot;.</li>
<li>This will be faster than using the browser to download a single XML file and it has the advantage of using multiprocessing in the next step as the script generates XML files with 100 titles in each document. The command also works with blocks of 500 PubMed-IDs, but not 1000.</li>
</ul>
</blockquote>
</li>
</ul>
</blockquote>
</li>
<li><p class="first">It is also possible to download the whole PubMed via license:</p>
<blockquote>
<ul class="simple">
<li><a class="reference external" href="http://www.nlm.nih.gov/databases/journal.html">http://www.nlm.nih.gov/databases/journal.html</a></li>
<li><a class="reference external" href="http://www.nlm.nih.gov/bsd/licensee/2015_stats/baseline_doc.html">http://www.nlm.nih.gov/bsd/licensee/2015_stats/baseline_doc.html</a></li>
</ul>
</blockquote>
</li>
<li><p class="first">You can insert the XML files in an extra directory in your project folder &quot;data&quot; or somewhere else. The only important thing is to use only one type of topic or XML files in one folder, because the programme will insert all files of type XML in a directory into the database given (next step).</p>
</li>
</ul>
</div>
<div class="section" id="installation">
<h1>Installation</h1>
<div class="section" id="operating-system">
<h2>Operating System</h2>
<ul class="simple">
<li>PubMed2Go was tested on Ubuntu and Fedora.</li>
</ul>
<div class="section" id="ubuntu">
<h3>Ubuntu</h3>
<ul>
<li><p class="first">These are the packages that need to be installed to use PubMed2Go:</p>
<blockquote>
<ul>
<li><p class="first">python&gt;=2.7.3</p>
</li>
<li><p class="first">postgresql&gt;=8.4</p>
</li>
<li><p class="first">python-xappy&gt;=0.5</p>
</li>
<li><p class="first">python-xapian&gt;=1.2.8</p>
</li>
<li><p class="first">python-sqlalchemy&gt;=0.9.7</p>
</li>
<li><p class="first">python-psycopg2&gt;=2.4.5 (dependency from SQLAlchemy)</p>
</li>
<li><p class="first">To install, the following command can be used in the Ubuntu terminal:</p>
<blockquote>
<ul class="simple">
<li>&quot;sudo apt-get install&quot;</li>
</ul>
</blockquote>
</li>
</ul>
</blockquote>
</li>
<li><p class="first">If you use an older Ubuntu version, you can use &quot;pip&quot; to upgrade your package versions specifically for your user name, e.g.:</p>
<blockquote>
<ul class="simple">
<li>&quot;sudo pip install sqlalchemy --upgrade&quot;</li>
</ul>
</blockquote>
</li>
</ul>
</div>
<div class="section" id="fedora">
<h3>Fedora</h3>
<ul>
<li><p class="first">This section describes how to install required packages and how to adjust PostgreSQL settings in Fedora.</p>
</li>
<li><p class="first">To install the Fedora packages use the following command. It will install all required packages:</p>
<blockquote>
<ul class="simple">
<li>&quot;sudo -E yum install python python-xappy python-sqlalchemy python-psycopg2 postgresql postgresql-server postgresql-contrib&quot;</li>
</ul>
</blockquote>
</li>
<li><p class="first">To enable PostgreSQL in Fedora, use the following steps:</p>
<blockquote>
<ul>
<li><p class="first">&quot;sudo systemctl enable postgresql&quot;</p>
</li>
<li><p class="first">To start postgresql use the following command</p>
<blockquote>
<ul class="simple">
<li>&quot;sudo systemctl start postgresql&quot;</li>
</ul>
</blockquote>
</li>
<li><p class="first">To populate initial data, the following command is required:</p>
<blockquote>
<ul class="simple">
<li>&quot;journalctl -xn&quot;</li>
</ul>
</blockquote>
</li>
<li><p class="first">To initialise database, use the following command:</p>
<blockquote>
<ul class="simple">
<li>&quot;sudo postgresql-setup initdb&quot;</li>
</ul>
</blockquote>
</li>
<li><p class="first">To allow read access to postgres, SELinux should be modified. This can be done with the following command:</p>
<blockquote>
<ul class="simple">
<li>&quot;grep postgres /var/log/audit/audit.log | audit2allow -M mypol&quot;</li>
</ul>
</blockquote>
</li>
<li><p class="first">Then you can do this:</p>
<blockquote>
<ul class="simple">
<li>&quot;sudo semodule -i mypol.pp&quot;</li>
</ul>
</blockquote>
</li>
<li><p class="first">Append this line in the file &quot;pg_hba.conf&quot; (default location: &quot;/var/lib/pgsql/data/pg_hba.conf&quot;):</p>
<blockquote>
<ul class="simple">
<li>&quot;host    all             all             0.0.0.0         0.0.0.0         trust&quot;</li>
<li>If &quot;trust&quot; is used instead of &quot;ident&quot;, you are allowed to use a password. &quot;0.0.0.0&quot; means that all machines are allowed to login. That means, if you want to customise which server has to reach the database, you can control it here.</li>
</ul>
</blockquote>
</li>
</ul>
</blockquote>
</li>
</ul>
</div>
</div>
<div class="section" id="creation-of-postgresql-superuser">
<h2>Creation of PostgreSQL superuser</h2>
<ul>
<li><p class="first">If there is not yet a superuser for the PostgreSQL database, create one with the name of your local account</p>
<blockquote>
<ul>
<li><p class="first">&quot;sudo -u postgres createuser --superuser &lt;user_name&gt;&quot;</p>
</li>
<li><p class="first">&quot;sudo -u &lt;user_name&gt; psql template1&quot;</p>
<blockquote>
<ul class="simple">
<li>\password &lt;press enter, type in password, and press enter, again&gt;</li>
<li>\q</li>
</ul>
</blockquote>
</li>
</ul>
</blockquote>
</li>
<li><p class="first">Now, you can connect to the standard PostgreSQL database &quot;postgres&quot; with PGAdmin3 or via command-line:</p>
<blockquote>
<ul class="simple">
<li>&quot;psql -h localhost -d postgres -U &lt;user_name&gt;&quot;</li>
</ul>
</blockquote>
</li>
</ul>
</div>
<div class="section" id="installation-with-docker">
<h2>Installation with Docker</h2>
<ul>
<li><p class="first">Docker is similar to a virtual machine,  but it is easier to deploy and more efficient. It was tested in Ubuntu and Windows.</p>
</li>
<li><p class="first">You can use the PubMed2Go image to create a PostgreSQL relational database and a Xapian full text index without installing the packages mentioned above in basically two steps.</p>
</li>
<li><p class="first">Install Docker - it was tested on Ubuntu (64-bit required):</p>
<blockquote>
<ul>
<li><p class="first"><a class="reference external" href="https://docs.docker.com/installation/ubuntulinux/">https://docs.docker.com/installation/ubuntulinux/</a></p>
</li>
<li><p class="first">There are many different operating systems supported:</p>
<blockquote>
<ul class="simple">
<li><a class="reference external" href="https://docs.docker.com/installation/#installation">https://docs.docker.com/installation/#installation</a></li>
</ul>
</blockquote>
</li>
</ul>
</blockquote>
</li>
<li><p class="first">Run Docker with the PubMed2Go image:</p>
<blockquote>
<ul>
<li><p class="first">Create a folder on your local disk with a name of your choice.</p>
</li>
<li><p class="first">Go into that folder and create a directory &quot;import_data&quot;.</p>
</li>
<li><p class="first">Copy you XML files downloaded from PubMed into the directory &quot;import_data&quot;.</p>
</li>
<li><p class="first">Open a terminal and type in this command:</p>
<blockquote>
<ul>
<li><p class="first">&quot;sudo docker run -d -v /home/&lt;user_name&gt;/&lt;folder_of_your_choice&gt;/:/export/ -p 9999:5432 bgruening/pubmed2go&quot;</p>
</li>
<li><p class="first">This will create the PostgreSQL folder as well as the full text index database folder within the &lt;folder_of_your_choice&gt;.</p>
</li>
<li><p class="first">You can see that Docker is running by typing in &quot;sudo docker ps&quot;. This will show a randomly generated name for your process.</p>
</li>
<li><p class="first">Stopping Docker is possible by doing &quot;sudo docker stop &lt;name&gt;&quot;.</p>
</li>
<li><p class="first">Docker maps your PosgreSQL port &quot;5432&quot; to the port &quot;9999&quot;. Now, you can connect to your database with PGAdmin via &quot;localhost&quot;, port &quot;9999&quot; and user &quot;parser&quot; with password &quot;parser&quot;. If you want to connect via command-line, use this command:</p>
<blockquote>
<ul class="simple">
<li>&quot;psql -h localhost -U parser -p 9999 -d pubmed&quot;</li>
</ul>
</blockquote>
</li>
<li><p class="first">If you have created another folder with a name &lt;folder_of_your_choice&gt; and the directory &quot;import_data&quot;, you can create another database on port &quot;9998&quot; and another full text index with different data there:</p>
<blockquote>
<ul class="simple">
<li>&quot;sudo docker run -d -v /home/&lt;user_name&gt;/&lt;folder_of_your_choice&gt;/:/export/ -p 9998:5432 bgruening/pubmed2go&quot;</li>
</ul>
</blockquote>
</li>
<li><p class="first">In case of replacing or creating a database on a port that is already used, delete the complete directory &lt;folder_of_your_choice&gt; and repeat the configuration steps.</p>
</li>
</ul>
</blockquote>
</li>
</ul>
</blockquote>
</li>
<li><p class="first">You can connect to PostgreSQL and Xapian with the programming language of your choice or follow the Python examples given in this documentation. If you want to develop your own text mining pipelines based on your data set of choice, you will have to install the required libraries on your operating system.</p>
</li>
<li><p class="first">This also means that you need a default PostgreSQL installation on your operating system. Restart a closed Docker session on port &quot;9999&quot; with the command:</p>
<blockquote>
<ul class="simple">
<li>&quot;sudo docker run -d -v /home/&lt;user_name&gt;/&lt;folder_of_your_choice&gt;/:/export/ -p 9999:5432 bgruening/pubmed2go&quot;</li>
</ul>
</blockquote>
</li>
<li><p class="first">It is not recommended to run the PubMed2Go examples or to develop new scripts within the Docker container. If you want to modify the image, use the Docker documentation and this repository:</p>
<blockquote>
<ul class="simple">
<li><a class="reference external" href="https://github.com/bgruening/docker-recipes/tree/master/pubmed2go">https://github.com/bgruening/docker-recipes/tree/master/pubmed2go</a></li>
</ul>
</blockquote>
</li>
<li><p class="first">If you want to try the examples given in the sections 5 to 8, copy the Xapian directory from the &lt;folder_of_your_choice&gt; into the folder &quot;PubMed2Go/full_text_index/xapian/&quot; from &quot;<a class="reference external" href="https://github.com/KerstenDoering/PubMed2Go">https://github.com/KerstenDoering/PubMed2Go</a>&quot; and run the Docker container in background. In case of using Docker, you can completely skip section 4.</p>
</li>
</ul>
</div>
</div>
<div class="section" id="build-up-a-relational-database-in-postgresql">
<h1>Build up a Relational Database in PostgreSQL</h1>
<ul>
<li><p class="first">Open a Terminal and type in:</p>
<blockquote>
<ul class="simple">
<li>&quot;psql template1&quot;</li>
</ul>
</blockquote>
</li>
<li><p class="first">Enter the following commands into psql prompt to create a database, the schema &quot;pubmed&quot;, and a standard user &quot;parser&quot;. It is important to write the user &quot;parser&quot; in single quotes in the creation step:</p>
<blockquote>
<ul class="simple">
<li>CREATE USER parser WITH PASSWORD 'parser';</li>
<li>CREATE DATABASE pancreatic_cancer_db;</li>
<li>GRANT ALL PRIVILEGES ON DATABASE pancreatic_cancer_db to parser;</li>
<li>\q</li>
</ul>
</blockquote>
</li>
<li><p class="first">Now you can create a schema &quot;pubmed&quot; as user &quot;parser&quot;. You will be asked to enter your password &quot;parser&quot; here:</p>
<blockquote>
<ul class="simple">
<li>&quot;psql -h localhost -d pancreatic_cancer_db -U parser -f create_schema.sql&quot;</li>
</ul>
</blockquote>
</li>
<li><p class="first">If you want to use another database name, just change &quot;pancreatic_cancer_db&quot; in these commands and provide this name in all other scripts by choosing the right parameter.</p>
</li>
<li><p class="first">It is recommended to use the name &quot;parser&quot; with password &quot;parser&quot; and the schema &quot;pubmed&quot;, because this is hard coded in &quot;PubMedDB.py&quot; and &quot;PubMedParser.py&quot;</p>
</li>
<li><p class="first">Create the tables in your database schema &quot;pubmed&quot; like this:</p>
<blockquote>
<ul class="simple">
<li>Use the command &quot;python PubMedDB.py -d pancreatic_cancer_db&quot; in your terminal. There are no other parameters that can be set.</li>
</ul>
</blockquote>
</li>
<li><p class="first">Load the data from PubMed into your PostgreSQL database:</p>
<blockquote>
<ul>
<li><p class="first">You can check &quot;python PubMedParser.py -h&quot; to get a help screen with all adjustable parameters. If you want to use the defaults, you can simply type in &quot;python PubMedParser.py&quot;.</p>
<blockquote>
<ul class="simple">
<li>By default, previously in PostgreSQL inserted data will be deleted before loading the new XML files into the database. That means you just have to call &quot;python PubMedParser.py&quot;, again in case you want to load new data into your already created database.</li>
<li>If you do not want to delete, but only add XML files to the data that is already inside your PostgreSQL database, use parameter &quot;-c&quot;.</li>
<li>The default database name is &quot;pancreatic_cancer_db&quot; and the default number of processors is 2. For changing, use parameters &quot;-d&quot; and &quot;-p&quot;.</li>
<li>If you want to process only part of your files, use the parameters &quot;-s&quot; and &quot;-e&quot; with numbers referring to your alphabetically sorted files, e.g. &quot;-s 0 -e 20&quot; for the first 20 XML files in the directory.</li>
</ul>
</blockquote>
</li>
<li><p class="first">It is important that you only type in the name of the folder containing all XML files with parameter &quot;-i&quot;, but not the name of the file(s). You do not need to type in the absolute path. Suppose, you have saved your XML file(s) in the directory &quot;data/pancreatic_cancer&quot;, use this command to run it with 3 processors and the database &quot;pancreatic_cancer_db&quot;:</p>
<blockquote>
<ul class="simple">
<li>&quot;python PubMedParser.py -i data/pancreatic_cancer/ -d pancreatic_cancer_db -p 3&quot;</li>
</ul>
</blockquote>
</li>
<li><p class="first">If you receive an error concerning too many database connections, make sure that you use the latest version of SQLAlchemy. In earlier versions, sometimes the database connections were closed by the programme, but still remained open for some seconds, preventing the new programme to open a new connection. You can also increase the number of possible connections to your PostgreSQL server that can be opened (Ubuntu: &quot;max_connections = &lt;type in number&gt;&quot; in &quot;/etc/postgresql/&lt;version number&gt;/main/postgresql.conf&quot;).</p>
</li>
<li><p class="first">For one file with around 272 MB this takes around 10 min (only one processor can be used). For the same amount of data split into files with only 100 PubMed-IDs (use &quot;generate_efetch.py&quot;) it takes around 4 min with 3 processors (2,83 GHz and 8 GB RAM).</p>
</li>
</ul>
</blockquote>
</li>
<li><p class="first">Now, a schema &quot;pubmed&quot; exists in your database &quot;pancreatic_cancer_db&quot; that contains all abstracts, titles, authors, etc. More information will be given in section 5, containing SQL queries and small programming examples.</p>
</li>
<li><p class="first">The schema is described in the file &quot;documentation/PostgreSQL_database_schema.html&quot; which was generated with DbSchema (<a class="reference external" href="http://www.dbschema.com/download.html">http://www.dbschema.com/download.html</a>).</p>
</li>
</ul>
</div>
<div class="section" id="build-up-a-full-text-index-with-xapian-and-search-it">
<h1>Build up a Full Text Index with Xapian and Search It</h1>
<ul>
<li><p class="first">The results from this section can be found in &quot;full_text_index/results/results_from_documentation/&quot;.</p>
</li>
<li><p class="first">Change into the directory &quot;full_text_index&quot; in your terminal.</p>
</li>
<li><p class="first">Create two directories, &quot;xapian&quot; and &quot;results&quot;, if they do not yet exist.</p>
</li>
<li><p class="first">Type in &quot;python RunXapian.py -h&quot; to get a help screen with all adjustable parameters.</p>
</li>
<li><p class="first">If you use all default values from this documentation, you will receive results in &quot;results/results.csv&quot; with &quot;python RunXapian.py -x&quot;.</p>
<blockquote>
<ul>
<li><p class="first">If you want to use the PostgreSQL database generated with Docker, change the port in the script &quot;Article.py&quot; in line 21 from &quot;5432&quot; to &quot;9999&quot; (the port you selected in Docker) and use the parameter &quot;-d&quot; with the database &quot;pubmed&quot; for the script &quot;RunXapian.py&quot;.</p>
</li>
<li><p class="first">This command indexes all titles, abstracts, keywords, MeSH terms and substances from year 1809 to 2015, downloaded as XML files from PubMed (as described in section &quot;Download a Data Set&quot;).</p>
<blockquote>
<ul class="simple">
<li>There are no abstracts with a publication date before 1809:</li>
<li><a class="reference external" href="http://www.nlm.nih.gov/bsd/licensee/2015_stats/baseline_med_filecount.html">http://www.nlm.nih.gov/bsd/licensee/2015_stats/baseline_med_filecount.html</a></li>
</ul>
</blockquote>
</li>
<li><p class="first">After completing the step of generating the full text index, the programme searches it with the synonyms given in &quot;synonyms/pancreatic_cancer.txt&quot;.</p>
<blockquote>
<ul class="simple">
<li>This file contains manually chosen names of drugs, genes, proteins, and diseases related to pancreatic cancer.</li>
<li>You can write your customized synonyms in a text file and can be passed to the Xapian while running RunXapian.py with the parameter &quot;-p&quot;.</li>
</ul>
</blockquote>
</li>
<li><p class="first">The output in the command-line shows how many PubMed-IDs are indexed (23258) and how many synonyms are searched (86).</p>
</li>
<li><p class="first">This takes around 2-3 min on a 2,83 GHz machine with 8 GB RAM.</p>
</li>
<li><p class="first">You can also select single years for indexing and searching.</p>
</li>
<li><p class="first">If you just want to index your XML files, type in &quot;python RunXapian.py -x -f&quot;. (Parameter &quot;-f&quot; turns off the search function of the programme, default is &quot;True&quot;.)</p>
</li>
<li><p class="first">If you just want to search your synonyms, type in &quot;python RunXapian.py&quot; (Parameter &quot;-x&quot; turns on the indexing step, default is &quot;False&quot;.)</p>
</li>
<li><p class="first">The default location for your full text index database folder is &quot;PubMed2Go/full_text_index/xapian/&lt;xapian2015&gt;&quot;. You can change this location by using the parameter &quot;-p&quot;.</p>
</li>
</ul>
</blockquote>
</li>
<li><p class="first">For the given example, 10392 lines were generated in &quot;results.csv&quot;. Run &quot;python summary.py&quot; to get two CSV files in directory &quot;results&quot;. If you have chosen another filename as output from &quot;RunXapian.py&quot;, you can do &quot;python summary.py -f &lt;name_of_input_file.csv&gt;&quot;:</p>
<blockquote>
<ul>
<li><p class="first">Drug synonyms were taken from DrugBank using the exact search query &quot;pancreatic cancer&quot;:</p>
<blockquote>
<ul class="simple">
<li><a class="reference external" href="http://www.drugbank.ca/">http://www.drugbank.ca/</a></li>
</ul>
</blockquote>
</li>
<li><p class="first">Protein and gene synonyms have been extracted manually from OMIM also performing an exact search:</p>
<blockquote>
<ul class="simple">
<li><a class="reference external" href="http://omim.org/entry/260350?search=%22pancreatic%20cancer%22">http://omim.org/entry/260350?search=%22pancreatic%20cancer%22</a></li>
</ul>
</blockquote>
</li>
<li><p class="first">Diseases related to pancreatic cancer have been taken the text given on OMIM, too.</p>
</li>
<li><p class="first">&quot;counts_results.csv&quot; shows how many synonyms were found (descending - 64 lines, meaning 64 from a total of 86 search terms). The alternative input filename will be &quot;counts_&lt;input_file.csv&gt;&quot;.</p>
<blockquote>
<ul class="simple">
<li>Taking into account the drugs, gemcitabine shows the most hits (2907). Erlotinib was found in 311 publications. Other approved drugs like WF10 and hydroxocobalamin were not found. Many investigational drugs were found 1-10 times: R115777, G17DT, hedgehog pathway inhibitor, imexon, GV1001, RP101, MGI-114, and PX-12. No other substances given on DrugBank were identified in this data set.</li>
<li>Pancreatic ductal adenocarcinoma is the most common type of pancreatic cancer ( <a class="reference external" href="http://www.cancer.gov/aboutnci/budget_planning_leg/plan-2013/profiles/pancreatic">http://www.cancer.gov/aboutnci/budget_planning_leg/plan-2013/profiles/pancreatic</a> ), which is shown by the 1598 hits. The tumor suppressor protein p53 was found 660 times, but also associated genes like KRAS, SMAD4, BRCA2, mTOR and CDKN2A were found (138-424 times). Many other genes were identified with a number below 10 hits and can be further analysed in &quot;pmids_results.csv&quot;.</li>
<li>Associated diseases like breast cancer, colon cancer, ovarian cancer and diabetes were found 255-919 times.</li>
</ul>
</blockquote>
</li>
<li><p class="first">&quot;pmids_results.csv&quot; shows which synonyms co-occur in the same abstract or title, sorted by PubMed-IDs (7500 lines). In case of an alternative input filename, there will be the resulting file &quot;pmids_&lt;input_file.csv&gt;&quot;.</p>
</li>
</ul>
</blockquote>
</li>
<li><p class="first">In case, you want to index the whole PubMed, it can be useful to index blocks of years or every year as a single directory. Like this, it is possible to use multiprocessing and decrease RAM usage. Just run the programme in different shells or on different machines and copy all resulting index folders to the same main directory. The tool &quot;xapian-compact&quot; summarises all generated directories to one full text index:</p>
<blockquote>
<ul class="simple">
<li><a class="reference external" href="http://xapian.org/docs/admin_notes.html#merging-databases">http://xapian.org/docs/admin_notes.html#merging-databases</a></li>
<li>xapian-compact -m  &lt;all input directories to be compressed, separated by space&gt; &lt;name of outcoming folder with complete database&gt;</li>
</ul>
</blockquote>
</li>
</ul>
</div>
<div class="section" id="examples-for-using-full-text-search-and-selecting-data-from-postgresql">
<h1>Examples for Using Full Text Search and Selecting Data from PostgreSQL</h1>
<div class="section" id="xapian">
<h2>Xapian</h2>
<ul>
<li><p class="first">Use the following scripts to work with the functions OR, AND, NEAR, ADJ, NOT, and phrase search in Xapian and have a look at the HTML output files. As the number of PubMed-IDs increases continuously, the resulting numbers in this documentation can be seen as a reference point for the given query &quot;pancreatic cancer&quot;. Having a look at these scripts as well as &quot;RunXapian.py&quot; can be useful to build your own modified queries. There is also a small note in &quot;full_text_index/xapian/readme.txt&quot;.</p>
<blockquote>
<ul>
<li><p class="first">&quot;python search_title.py&quot; shows that only a few lines of code are required to search only publication titles. This can be important as searching especially in publication titles puts more emphasis on the queried synonyms.</p>
<blockquote>
<ul class="simple">
<li>While &quot;RunXapian.py&quot; searches only the exact phrase &quot;pancreatic cancer&quot;, &quot;search_title.py&quot; searches for the stem &quot;pancreat&quot; and also finds the word &quot;pancreatitis&quot;. You can change the querystring in the script search_title.py for phrase you want to search.</li>
<li>It generates &quot;Xapian_query_results.html&quot; which shows the first 1000 of 18085 titles. Like this, many associated words are shown, e.g. &quot;pancreatic ductal adenocarcinoma&quot;, &quot;pancreatic juice&quot;, or &quot;pancreatic diseases&quot;.</li>
</ul>
</blockquote>
</li>
<li><p class="first">To further specify your search, you can query titles containing &quot;pancreatic cancer&quot; and the drug &quot;erlotinib&quot; with &quot;python search_near_title.py&quot;.</p>
<blockquote>
<ul class="simple">
<li>This generates 38 results in &quot;Xapian_query_results_NEAR.html&quot;.</li>
<li>In this case &quot;NEAR/5&quot; is used as a Xapian function. In this case, a maximum of 4 words is allowed to be between the two search terms.</li>
<li>An alternative would be the query with &quot;ADJ/5&quot;, which reduces the number of 38 hits to 4 hits, because with this function, the order of search terms is fixed.</li>
<li>Here, the exact search is performed, again.</li>
<li>These querries can be changed in the script search_near_title.py edit the inputs to the variables querystring1 and querystring2.</li>
</ul>
</blockquote>
</li>
<li><p class="first">As it was done in &quot;RunXapian.py&quot; different index fields can be searched.</p>
<blockquote>
<ul class="simple">
<li>&quot;python search_title_or_text.py&quot; searches documents in which the drug &quot;R115777&quot; occurs in the title or the text.</li>
<li>As shown in &quot;counts_results.csv&quot;, only 10 hits can be found. The matching titles and abstracts can be seen in &quot;Xapian_query_results_OR.html&quot;.</li>
</ul>
</blockquote>
</li>
<li><p class="first">The script &quot;python search_not_title_or_text.py&quot; specifies the query to documents not containing the terms &quot;colon&quot;, &quot;lung&quot;, or &quot;ovarian&quot;, but the word &quot;pancreatic&quot;.</p>
<blockquote>
<ul class="simple">
<li>This reduces the number of results to 9 hits, as no publications are considered that contain these other types of cancer.</li>
<li>The result is shown in &quot;Xapian_query_results_NOT.html&quot;.</li>
</ul>
</blockquote>
</li>
<li><p class="first">In this way, different search queries can be combined with a few lines of code.</p>
</li>
</ul>
</blockquote>
</li>
</ul>
</div>
<div class="section" id="postgresql">
<h2>PostgreSQL</h2>
<ul>
<li><p class="first">Type in these SQL queries in PGAdmin3 or in the PostgreSQL shell to get familiar with the schema &quot;pubmed&quot;:</p>
<blockquote>
<ul>
<li><p class="first">Find all substances related to pancreatic cancer, pancreatitis, etc.</p>
<blockquote>
<ul class="simple">
<li>select * from pubmed.tbl_chemical where lower(name_of_substance) LIKE 'pancreati%'; -- 180 lines</li>
</ul>
</blockquote>
</li>
<li><p class="first">Find all MeSH terms with the substring &quot;ancreat&quot; and prefixes as well as suffixes.</p>
<blockquote>
<ul class="simple">
<li>select distinct on (descriptor_name) * from pubmed.tbl_mesh_heading where lower(descriptor_name) LIKE '%ancreat%'; -- 29 lines</li>
</ul>
</blockquote>
</li>
<li><p class="first">What is the number of published titles in our database?</p>
<blockquote>
<ul class="simple">
<li>select count(*) from pubmed.tbl_medline_citation; -- 23258</li>
</ul>
</blockquote>
</li>
<li><p class="first">How many publications contain an abstract?</p>
<blockquote>
<ul class="simple">
<li>select count(*) from pubmed.tbl_abstract; -- 21387</li>
</ul>
</blockquote>
</li>
<li><p class="first">Show me all different journals and abbreviations referring to our topic.</p>
<blockquote>
<ul class="simple">
<li>select distinct on (title, iso_abbreviation) title, iso_abbreviation from pubmed.tbl_journal; -- 2209 lines</li>
</ul>
</blockquote>
</li>
<li><p class="first">What is the number of publications since 1990?</p>
<blockquote>
<ul class="simple">
<li>select count(*) from pubmed.tbl_journal where pub_date_year &lt;=2000 and pub_date_year &gt;=1990; -- between 1990 and 2000: 3736 publications</li>
<li>select count(*) from pubmed.tbl_journal where pub_date_year &lt;=2010 and pub_date_year &gt;2000; -- after 2000 until 2010: 9497 publications</li>
<li>select count(*) from pubmed.tbl_journal where pub_date_year &gt;2010; -- after 2010: 8461 publications</li>
</ul>
</blockquote>
</li>
<li><p class="first">What is the number of publications in USA referring to our topic?</p>
<blockquote>
<ul class="simple">
<li>select count(*) from pubmed.tbl_medline_journal_info where lower(country) = 'united states'; -- 10010 publications</li>
</ul>
</blockquote>
</li>
<li><p class="first">Take one of the first publications for the query &quot;pancreatic cancer&quot; in the browser on NCBI and check whether this author has other publications, e.g. Bobustuc et al., 2015.</p>
<blockquote>
<ul class="simple">
<li>select count (*) from pubmed.tbl_author where last_name = 'Bobustuc'; -- 2</li>
</ul>
</blockquote>
</li>
</ul>
</blockquote>
</li>
</ul>
</div>
<div class="section" id="postgresql-and-xapian">
<h2>PostgreSQL and Xapian</h2>
<ul>
<li><p class="first">The results of this subsection can be found in &quot;full_text_index/results/results_from_documentation/&quot;.</p>
</li>
<li><p class="first">Try &quot;python find_authors.py&quot; to see an example for processing a PostgreSQL query in Python. Use &quot;python find_authors.py -f &lt;output_filename&gt; -d &lt;name_of_database&gt;&quot; to specify the name of the output file and the database to connect to. &quot;python find_authors.py -h&quot; shows all adjustable parameters.</p>
<blockquote>
<ul>
<li><p class="first">Considering the output file &quot;results/authors.csv&quot;, Ralph H. Hruban has published the most articles with a number of 258 PubMed-IDs.</p>
<blockquote>
<ul class="simple">
<li>The other authors and their number of publications can be found in descending order.</li>
</ul>
</blockquote>
</li>
<li><p class="first">You can check the amount of publications from similiarly written author names in PGAdmin3 and then Helmut Friess is shown as the one with the most publications:</p>
<blockquote>
<ul class="simple">
<li>select distinct on(fk_pmid) * from pubmed.tbl_author where last_name = 'Friess' and (fore_name = 'H' or fore_name = 'Helmut') order by fk_pmid; -- 375</li>
<li>select distinct on(fk_pmid) * from pubmed.tbl_author where last_name = 'Büchler' and (fore_name = 'Markus W' or fore_name = 'M W') order by fk_pmid; -- 310</li>
<li>select distinct on(fk_pmid) * from pubmed.tbl_author where last_name = 'Hruban' and fore_name = 'Ralph H'; -- 258</li>
</ul>
</blockquote>
</li>
<li><p class="first">It is possible that an author name exists twice although different persons are meant. This is not considered here.</p>
</li>
<li><p class="first">There are examples in which you can only find a collective name:</p>
<blockquote>
<ul class="simple">
<li>select * from pubmed.tbl_author where last_name is NULL and fore_name is NULL; -- 273</li>
</ul>
</blockquote>
</li>
</ul>
</blockquote>
</li>
<li><p class="first">Based on this, it is possible to consider whether the author Helmut Friess has published something containing the query terms from the list in &quot;synonyms/pancreatic_cancer.txt&quot;:</p>
<blockquote>
<ul>
<li><p class="first">Type in &quot;python find_topics.py&quot;. You can try &quot;python find_topics.py -h&quot;, to see which parameters can be varied, e.g. if your input filename is not &quot;pmids_results.csv&quot; or if you want to specify your output filename, which default is &quot;pmids_results_from_author.csv&quot;.</p>
<blockquote>
<ul class="simple">
<li>127 publications were found for the given list of synonyms and this author.</li>
<li>The main research topic seems to be pancreatic ductal adenocarcinoma. This result can be compared with the outputs using other author names (hard coded in &quot;find_topics.py&quot;) and running &quot;find_topics.py&quot; with another filename, again.</li>
</ul>
</blockquote>
</li>
</ul>
</blockquote>
</li>
<li><p class="first">Next steps can be to select the abstracts that were identified with Xapian from PostgreSQL and to apply software for named entity recognition (section &quot;Examples for Using BioC and PubTator&quot;) or to visualise data (next section). There are many possibilities to develop customised pipelines, e.g. selecting sentences, applying part-of-speech tagging, and train machine learning models to extract semantic relationships.</p>
</li>
</ul>
</div>
</div>
<div class="section" id="examples-for-generating-plots">
<h1>Examples for Generating Plots</h1>
<ul class="simple">
<li>This section covers the generation of word clouds and pie charts, for which special software packages are needed.</li>
</ul>
<div class="section" id="word-cloud">
<h2>Word Cloud</h2>
<blockquote>
<ul>
<li><p class="first">The word clouds generated here are based on the modified Xapian full text version searching only PubMed titles and abstract texts. Therefore, the files &quot;RunXapian.py&quot; and &quot;SynonymParser.py&quot; as well as the folder &quot;synonyms&quot; need to be copied from the folder &quot;full_text_index&quot; to the folder &quot;full_text_index_title_text&quot;. The directories &quot;xapian&quot; and &quot;results&quot; have to be created, too. Afterwards, the command &quot;python RunXapian.py -x&quot; can be used, again. The numbers described in the last sections can differ slightly from the results generated here. The command &quot;python summary.py&quot; also has to executed.</p>
</li>
<li><p class="first">At first, the list of the 50 most frequently occurring words that were generated with &quot;python summary.py&quot; needs to be extracted in logarithmic scale to visualise the search terms appropriately. In the directory &quot;PubMed2Go/plots/word_cloud&quot;, run the script &quot;get_search_terms_log.py&quot; to get the output file &quot;counts_search_terms_log.csv&quot;. The highest frequency is shown by the small molecule gemcitabine. The parameter &quot;-h&quot; shows available parameters.</p>
</li>
<li><p class="first">The second step in this example is to find the 50 most frequently co-occurring words in texts that contain the search term gemcitabine. This can be done by running the command &quot;python generate_surrounding_words_log.py&quot;. The stop word list that is used by this script was referenced by Hettne et al. [A dictionary to identify small molecules and drugs in free text. Bioinformatics. 2009 Nov 15;25(22):2983-91. doi: 10.1093/bioinformatics/btp535.]. It is provided in the folder &quot;blacklist&quot;. Have a look at the links given in &quot;stop_words.txt&quot;. These stop words were used to filter out terms with a very high frequency that have no substantial meaning for the content analysed. The numbers in the ouput file &quot;counts_surrounding_words_log.csv&quot; are given in logarithmic scale, too.</p>
</li>
<li><p class="first">The plot can be genrated with the package &quot;PyTagCloud&quot;. Please, follow the installation instructions on this GitHub page:</p>
<blockquote>
<ul class="simple">
<li><a class="reference external" href="https://github.com/atizo/PyTagCloud">https://github.com/atizo/PyTagCloud</a></li>
</ul>
</blockquote>
</li>
<li><p class="first">For the first plot given here, use the command &quot;python create_word_cloud.py -i counts_search_terms_log.csv -o cloud_search_terms.png&quot;. Adjust the parameters in the script create_word_cloud.py in the line create_tag_image, change parameter in size method of the output file to get the better cloud image.</p>
</li>
</ul>
<img alt="cloud_search_terms_700_w.png" src="cloud_search_terms_700_w.png" />
<ul>
<li><p class="first">For the next figure, run &quot;python create_word_cloud.py -i counts_surrounding_words_log.csv -o cloud_surrounding_words.png&quot;. Here also adjust the parameters in the script create_word_cloud.py in the line create_tag_image, change parameter in size method of the output file to get the better cloud image.</p>
<blockquote>
<img alt="cloud_surrounding_words_500_w.png" src="cloud_surrounding_words_500_w.png" />
</blockquote>
</li>
<li><p class="first">The word clouds will look different every time the script is used.</p>
</li>
</ul>
</blockquote>
</div>
<div class="section" id="pie-chart">
<h2>Pie Chart</h2>
<blockquote>
<ul class="simple">
<li>In this subsection, the library &quot;matplotlib&quot; is needed to generate a pie chart. You can install it with the command &quot;sudo apt-get install python-matplotlib&quot; in ubuntu or with command &quot;dnf install python-matplotlib&quot; from Fedora 22 or with command &quot;yum install python-matplotlib&quot; in Fedora &lt; 22.</li>
<li>By running &quot;python pie_chart_countries.py&quot;, the picture &quot;pie_chart_countries_publications.png&quot; is produced from the input file &quot;countries_pancreatic_cancer.csv&quot;.</li>
<li>To get the CSV file, you need to connect to your database, e.g. with &quot;psql -h localhost -d pancreatic_cancer_db -U parser&quot; and type in &quot;\COPY (SELECT fk_pmid, LOWER(country) FROM pubmed.tbl_medline_journal_info WHERE country IS NOT NULL ORDER BY country ASC) TO 'countries_pancreatic_cancer.csv' DELIMITER ','&quot;.</li>
<li>The script calculates the percentages of country names in which the journals given in the PostgreSQL database are published. Fractions below 2 % are summarised to &quot;Others&quot;.</li>
<li>The plot was inspired by an example given in the Matplotlib documentation (<a class="reference external" href="http://matplotlib.org/examples/pie_and_polar_charts/pie_demo_features.html">http://matplotlib.org/examples/pie_and_polar_charts/pie_demo_features.html</a>).</li>
</ul>
<img alt="pie_chart_countries_publications_700_w.png" src="pie_chart_countries_publications_700_w.png" />
</blockquote>
</div>
<div class="section" id="bar-chart">
<h2>Bar Chart</h2>
<blockquote>
<ul class="simple">
<li>In this subsection, the library &quot;matplotlib&quot; is needed, too.</li>
<li>Three timelines for the publications of the genes KRAS, BRCA2, and CDKN2A are shown in one bar chart.</li>
<li>Running &quot;create_bar_chart.py -p&quot; generates the figure &quot;KRAS_BRCA2_CDKN2A_pubmed.png&quot;. The year 2015 cannot be considered as a complete year. Therefore, it is removed by this script before plotting.</li>
<li>The CSV files processed by this script can be downloaded from PubMed by clicking on the bar chart appearing on <a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed">http://www.ncbi.nlm.nih.gov/pubmed</a> after entering the query (15th June 2015). The title lines in these CSV files were removed manually.</li>
<li>All CSV files used in this subsection are written in comma-separated format.</li>
</ul>
<img alt="../plots/bar_chart/KRAS_BRCA2_CDKN2A_pubmed.png" src="../plots/bar_chart/KRAS_BRCA2_CDKN2A_pubmed.png" />
<ul class="simple">
<li>Running &quot;python get_years.py&quot; generates the same kind of CSV files as provided by the browser search, but it uses the pancreatic cancer data set from this documentation by sending a query to the PubMed2Go PostgreSQL database.</li>
<li>Based on this, &quot;create_bar_chart.py&quot; without the parameter &quot;-p&quot; generates the bar chart &quot;KRAS_BRCA2_CDKN2A.png&quot;.</li>
</ul>
<img alt="../plots/bar_chart/KRAS_BRCA2_CDKN2A.png" src="../plots/bar_chart/KRAS_BRCA2_CDKN2A.png" />
<ul class="simple">
<li>The slopes of the BRCA2 and CDKN2A timelines are rather low compared to KRAS, but start much earlier in both plots. There is even a decrease shown for the last three years in the pancreatic cancer data set. The timeline of the gene KRAS shows an exponential growth. One reason for this is its role in the regulation of cell proliferation [Small molecule inhibition of the KRAS-PDEδ interaction impairs oncogenic KRAS signalling. Zimmermann et al. Nature. 2013 May 30;497(7451):638-42. doi: 10.1038/nature12205. Epub 2013 May 22.].</li>
<li>The review on OMIM mentioned in section 5 (<a class="reference external" href="http://omim.org/entry/260350?search=%22pancreatic%20cancer%22">http://omim.org/entry/260350?search=%22pancreatic%20cancer%22</a>) provides more information with references showing why and how specific these genes are related to pancreatic cancer.</li>
</ul>
</blockquote>
</div>
</div>
<div class="section" id="examples-for-using-bioc-and-pubtator">
<h1>Examples for Using BioC and PubTator</h1>
<ul>
<li><p class="first">Follow the installation instructions on this GitHub page:</p>
<blockquote>
<ul class="simple">
<li><a class="reference external" href="https://github.com/2mh/PyBioC">https://github.com/2mh/PyBioC</a>.</li>
</ul>
</blockquote>
</li>
<li><p class="first">Copy the project folder &quot;PyBioC/src/bioc&quot; into your folder &quot;PubMed2Go/BioC_export&quot;. This PyBioC directory is needed for the script &quot;add_BioC_annotation.py&quot;, because it contains the Python source code for the BioC interface.</p>
</li>
<li><p class="first">The idea of BioC is to use a standardised XML format that can be shared by the community to add annotations to scientific texts. Therefore, these documents can be exchanged and modified by anybody who uses the BioC interface.</p>
</li>
<li><p class="first">The BioC XML format was introduced at BioCreative (<a class="reference external" href="http://www.biocreative.org/events/BCBioCuration2014/biocreative-text-mining-worksh">http://www.biocreative.org/events/BCBioCuration2014/biocreative-text-mining-worksh</a>) and also used at BioNLP (<a class="reference external" href="http://2013.bionlp-st.org/supporting-resources">http://2013.bionlp-st.org/supporting-resources</a>). The BioC project homepage contains several related software packages (<a class="reference external" href="http://bioc.sourceforge.net">http://bioc.sourceforge.net</a>).</p>
</li>
<li><p class="first">The file &quot;pmid_list.txt&quot; contains 21 PubMed-IDs that were taken from &quot;PubMed2Go/data/pubmed_result.txt&quot;. It is used as default by the script &quot;write_BioC_XML.py&quot;. If you do not have pubmed_result.txt file which was one of the alternative to get the dataset done in the section Download a Data set, you can get the ids by saving them to &quot;pmid_list.txt&quot; from the table tbl_abstract or/and tbl_medline_citation and also make sure you get some pubmed IDs from table tbl_mesh_heading so that the annotation works fine in the later example when you run the script add_BioC_annotation.py of the datbase in this case it is pancreatic_cancer_db.</p>
</li>
<li><p class="first">This script also uses the file &quot;BioC.dtd&quot;, which defines the structure of the XML file (taken from <a class="reference external" href="https://github.com/2mh/PyBioC">https://github.com/2mh/PyBioC</a>).  Additionally, the file &quot;Explanation.key&quot; describes the semantics used for the annotations. In this example, MeSH terms are added as annotation XML elements to the basic BioC XML structure.</p>
</li>
<li><p class="first">For the list of PubMed-IDs, the command &quot;python write_BioC_XML.py&quot; generates the BioC XML file &quot;text_BioC.xml&quot;.</p>
</li>
<li><p class="first">If you run &quot;python add_BioC_annotation.py&quot; with default values, it adds the MeSH terms provided by the PostgreSQL database to the file &quot;text_BioC.xml&quot;. It does not change the basic structure, but it writes annotation XML elements for the text elements given. The output file is called &quot;annotated_text_BioC.xml&quot;. This is an example for an XML file that did not contain annotation elements before.</p>
</li>
<li><p class="first">You can use the same script for adding the MeSH annotation elements to an XML document that already contains other annotation elements as shown in the next example.</p>
</li>
<li><p class="first">PubTator can be used as a webservice in several ways. There is a website highlighting entities like genes/proteins, chemical compounds, diseases, mutations, and species. This website can be used for text curation:</p>
<blockquote>
<ul class="simple">
<li><a class="reference external" href="http://www.ncbi.nlm.nih.gov/CBBresearch/Lu/Demo/PubTator">http://www.ncbi.nlm.nih.gov/CBBresearch/Lu/Demo/PubTator</a></li>
</ul>
</blockquote>
</li>
<li><p class="first">There is also a RESTful API:</p>
<blockquote>
<ul class="simple">
<li><a class="reference external" href="http://www.ncbi.nlm.nih.gov/CBBresearch/Lu/Demo/tmTools/curl.html">http://www.ncbi.nlm.nih.gov/CBBresearch/Lu/Demo/tmTools/curl.html</a></li>
</ul>
</blockquote>
</li>
<li><p class="first">How to use this API is shown by the following command executed via command-line:</p>
<blockquote>
<p>curl -H &quot;content-type:application/json&quot; http://www.ncbi.nlm.nih.gov/CBBresearch/Lu/Demo/RESTful/tmTool.cgi/Disease/1000475,1006519,1010707/BioC/ &gt; text_PubTator.xml</p>
<ul class="simple">
<li>All output files in this section refer to these three PubMed-IDs. The maximum number of PubMed-IDs to send to PubTator in this case was 21. This is the reason why &quot;pmid_list.txt&quot; contains exactly 21 PubMed-IDs.</li>
</ul>
</blockquote>
</li>
<li><p class="first">The script &quot;call_PubTator.py&quot; wraps this command with the Python module &quot;subprocess&quot; and downloads PubMed BioC XML annotated abstracts from PubTator.</p>
<blockquote>
<ul>
<li><p class="first">It contains the parameter &quot;-t&quot; (trigger) that selects the type of entity to be tagged (default: Disease) for a list of PubMed-IDs (default: pmid_list.txt).</p>
</li>
<li><p class="first">All parameters of this script can be shown with &quot;python call_PubTator.py -h&quot;.</p>
</li>
<li><p class="first">The default output file &quot;text_PubTator.xml&quot; also shows the MeSH IDs for the extracted diseases, e.g. &quot;pancreatic carcinoma&quot;, the first one in the example file:</p>
<blockquote>
<ul class="simple">
<li><a class="reference external" href="http://www.nlm.nih.gov/cgi/mesh/2011/MB_cgi?field=uid&amp;term=D010190">http://www.nlm.nih.gov/cgi/mesh/2011/MB_cgi?field=uid&amp;term=D010190</a></li>
</ul>
</blockquote>
</li>
<li><p class="first">PubTator returns two types of infon elements. Therefore, the line &quot;&lt;!ELEMENT annotation ( infon*, location*, text ) &gt;&quot; had to be changed to &quot;&lt;!ELEMENT annotation ( infon*, location*, text, infon* ) &gt;&quot;.</p>
</li>
</ul>
</blockquote>
</li>
<li><p class="first">If you want to add MeSH term annotations from the PostgreSQL to the file &quot;text_PubTator.xml&quot;, you can run the command &quot;python add_BioC_annotation.py -i text_PubTator.xml -o annotated_text_PubTator.xml&quot;.</p>
<blockquote>
<ul class="simple">
<li>MeSH terms refer to several types of entities. In this case, some of the MeSH terms will show duplicate disease annotation elements.</li>
</ul>
</blockquote>
</li>
<li><p class="first">All these entities can also be tagged in BioC XML format from plain text input via the single software packages referenced here below &quot;Quick Links&quot;:</p>
<blockquote>
<ul class="simple">
<li><a class="reference external" href="http://www.ncbi.nlm.nih.gov/CBBresearch/Lu/Demo/tmTools">http://www.ncbi.nlm.nih.gov/CBBresearch/Lu/Demo/tmTools</a></li>
</ul>
</blockquote>
</li>
<li><p class="first">They are described in the following PDF file as well as other software packages in chapter &quot;TRACK 1 (BioC: Interoperability)&quot;</p>
<blockquote>
<ul class="simple">
<li><a class="reference external" href="http://www.biocreative.org/media/store/files/2013/">http://www.biocreative.org/media/store/files/2013/</a><a href="#id1"><span class="problematic" id="id2">ProceedingsBioCreativeIV_vol1_</span></a>.pdf</li>
<li>There are also other webservices included as well as BioC natural language preprocessing pipelines in C++ and Java (<a class="reference external" href="http://bioc.sourceforge.net">http://bioc.sourceforge.net</a>).</li>
</ul>
</blockquote>
</li>
</ul>
</div>
<div class="section" id="contact">
<h1>Contact</h1>
<ul>
<li><p class="first">Please, write an e-mail, if you have questions, feedback, improvements, or new ideas:</p>
<blockquote>
<ul class="simple">
<li><a class="reference external" href="mailto:kersten.doering&#64;pharmazie.uni-freiburg.de">kersten.doering&#64;pharmazie.uni-freiburg.de</a></li>
</ul>
</blockquote>
</li>
<li><p class="first">If you are interested in related projects, visit our working group's homepage:</p>
<blockquote>
<ul class="simple">
<li><a class="reference external" href="http://www.pharmaceutical-bioinformatics.de">http://www.pharmaceutical-bioinformatics.de</a></li>
</ul>
</blockquote>
</li>
</ul>
<div class="section" id="license">
<h2>License</h2>
<ul class="simple">
<li>PubMed2Go is published with an ISC license given in &quot;license.txt&quot;.</li>
</ul>
</div>
</div>
<div class="system-messages section">
<h1>Docutils System Messages</h1>
<div class="system-message" id="id1">
<p class="system-message-title">System Message: ERROR/3 (<tt class="docutils">wiki.rst</tt>, line 609); <em><a href="#id2">backlink</a></em></p>
Unknown target name: &quot;proceedingsbiocreativeiv_vol1&quot;.</div>
</div>
</div>
</body>
</html>
